utils/db_utils.py

```diff
@@ -158,4 +159,87 @@ def add_user_stats(user_data):
             if last_call != -1:
                 user['last_call'] = arrow.get(last_call).format(time_format)
 
-    return user_data
\ No newline at end of file
+    return user_data
+
+def query_segments_crossing_endpoints(start_lat, start_long, end_lat, end_long, range_around_endpoints=400):
+    # data.loc only appears in analysis/recreated_location
+    query_start = {'data.loc': {'$near': SON([('$geometry', SON([('type', 'Point'), ('coordinates', [start_long, start_lat])])), ('$maxDistance', range_around_endpoints)])}}
+    query_end = {'data.loc': {'$near': SON([('$geometry', SON([('type', 'Point'), ('coordinates', [end_long, end_lat])])), ('$maxDistance', range_around_endpoints)])}}
```

Comment:
it seems like it would be better to use `emission/storage/timeseries/geoquery.py` with the timeseries in `emission/storage/timeseries`. That will make it easier to migrate to a different DB/storage solution if we want to later.

Note that this container is built on the e-mission-server base container and has access to the full e-mission-server codebase.

I do note that the current implementation of `emission/storage/timeseries/geoquery` only supports `$geoWithin` a `Polygon` or `MultiPolygon`, but it should be easy to extend. Let us know if you would like us to tackle that instead.

I'm also curious as to why you needed `SON` to get this to work - the current polygon based geoquery is used in a unit test with a simple dict, no `SON` and it seems to work.

Reply:
The `SON` lib was indeed unneeded, I removed it in commit https://github.com/e-mission/op-admin-dashboard/pull/61/commits/25e51792d786541ef547f821b8cc38f15567213d

For the geoquery extension, I believe I might need a hand. I'm running into issues with `count_documents` not supporting `$near` queries. (See https://www.mongodb.com/community/forums/t/why-the-countdocuments-doesnt-support-geonear/1330)
This limitation leads to an error here: https://github.com/e-mission/e-mission-server/blob/a8c7dadac63480745c88b8b1cc643df6fc70fa7b/emission/storage/timeseries/builtin_timeseries.py#L232 that I'm not sure how to handle correctly.
`pymongo.errors.OperationFailure: $geoNear, $near, and $nearSphere are not allowed in this context, full error: {'ok': 0.0, 'errmsg': '$geoNear, $near, and $nearSphere are not allowed in this context', 'code': 2, 'codeName': 'BadValue'}`

Reply:
@TTalex `$geoWithin` seems to work with `count_documents` (https://github.com/e-mission/e-mission-server/pull/936/commits/b6d90cfb232960a04cf93e475c5780ba9abc67a0). You should be able to use the `$center` operator as a drop-in replacement for your code. Although if you have time to investigate `leaflet.draw` (https://leaflet.github.io/Leaflet.draw/docs/leaflet-draw-latest.html) and its use in plotly, you could allow the user to draw the polygon at the start and end and then use `geoWithin` with a polygon.

Note that this likely strictly better - users can have a different radius for start and end, and they can draw their polygon to exclude intersections they don't want... But I don't know how easy that leaflet draw integration is.

Finally, please note that `$near` is less performant because it sorts the results and `$geoWithin` is recommended for better performance unless you need the sorting.
https://www.mongodb.com/docs/manual/reference/operator/query/near/

> If the order of the documents is not important to you, consider using the [$geoWithin](https://www.mongodb.com/docs/manual/reference/operator/query/geoWithin/#mongodb-query-op.-geoWithin) operator instead, as it returns unsorted results.

Reply:
Good insight, thanks for that.

Using draw was the right option, it's way more precise for users than circles, and it looks like it's working well with the existing Geoquery method.

I re-wrote most of the code to use this solution in commit https://github.com/e-mission/op-admin-dashboard/pull/61/commits/59e0c9810e68981593afef224a345f5c38732227

Moreover, dash-leaflet went into v1.x.x a few days ago, breaking some interactions with circles and offering a clean interaction with leflet.draw. So timing is great !